// @ionic-super-tabs/angular/esm2015/app-init.js
import * as tslib_1 from "tslib";
import { applyPolyfills, defineCustomElements } from '@ionic-super-tabs/core/loader';
let didInitialize = false;
export function appInit(doc, zone) {
    return function () {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            console.log('init');
            const win = doc.defaultView;
            console.log('Initializing');
            if (!win || didInitialize) {
                return;
            }
            didInitialize = true;
            const aelFn = '__zone_symbol__addEventListener' in doc.body
                ? '__zone_symbol__addEventListener'
                : 'addEventListener';
            yield applyPolyfills();
            yield defineCustomElements(win, {
                syncQueue: true,
                raf,
                jmp: (h) => zone.runOutsideAngular(h),
                ael(elm, eventName, cb, opts) {
                    elm[aelFn](eventName, cb, opts);
                },
                rel(elm, eventName, cb, opts) {
                    elm.removeEventListener(eventName, cb, opts);
                },
            });
        });
    };
}
;
export const raf = (h) => {
    if (typeof __zone_symbol__requestAnimationFrame === 'function') {
        return __zone_symbol__requestAnimationFrame(h);
    }
    if (typeof requestAnimationFrame === 'function') {
        return requestAnimationFrame(h);
    }
    return setTimeout(h);
};
